File: Word Search.cpp
class Solution {
public:
    bool dfs(vector<int> index, vector<vector<char>> &board, string word, int position, vector<vector<bool>> &vis ){          
        //base case
        if(position == word.size()){
            return true;
        }
        
        //recursive calls
        int nrow = board.size();
        int ncol = board[0].size();
        
        int r = index[0];
        int c = index[1];
        
        vis[r][c] = true;
        
        vector<int> dx = {0, 0, 1, -1};
        vector<int> dy = {1, -1, 0, 0};
        
        for(int i=0; i<4; i++){
            int row = r + dx[i];
            int col = c + dy[i];
            if(row>=0 && row<nrow && col>=0 && col<ncol){
                if(board[row][col]==word[position] && !vis[row][col]){
                    if(dfs({row, col}, board, word, position+1, vis) == true){
                        return true;
                    }
                    //backtracking
                    vis[row][col] = false;
                }
            }
        }
        
        return false;
        
    }
public:
    bool isWordExist(vector<vector<char>>& board, string word) {
        //find the index of the first character
        vector<int> index;
        bool done = false;
        int row = board.size();
        int col = board[0].size();
        vector<vector<bool>> visited(row, vector<bool>(col, false));
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(board[i][j]==word[0]){
                    //call dfs function form here
                    if(dfs({i, j}, board, word, 1, visited) == true){
                        return true;
                    }
                }
            }
        }
        
        return false;
    }
};

File: Word Search.java
class Solution{
    public boolean isWordExist(char[][] board, String word){
        // Code here
        int row = board.length;
        int col = board[0].length;
        int[][] visited = new int[row][col];
        int k=0;
        
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(find(board, word, i, j, row, col, k, visited))
                    return true;
            }
        }
        return false;
    }
    
    public boolean find(char[][] board, String word, int i, int j, int row, int col, int k, int[][] visited){
        if(k==word.length())
            return true;
            
        if(i<0 || i>=row || j<0 || j>=col || visited[i][j]==1 || board[i][j]!=word.charAt(k))
            return false;
            
        visited[i][j]=1; 
        boolean up = find(board, word, i-1, j, row, col, k+1, visited);
        boolean down = find(board, word, i+1, j, row, col, k+1, visited);
        boolean left = find(board, word, i, j-1, row, col, k+1, visited);
        boolean right = find(board, word, i, j+1, row, col, k+1, visited);
        visited[i][j]=0;
        
        return up || down || left || right;
                
    }
}
